### ISSUE TITLE
Migration: Helm 2 to Helm 3

### ISSUE THREAD

**Issue by hickeyma** [helm/helm#5892]
This feature is to cover migration from Helm 2 to Helm 3.

The key changes in Helm 3 which will directly affect Helm 2 are as follows:
- Release state storage. The changes includes the Kubernetes resource for storage and the release object metadata contained in the resource. Releases are now on a per user namespace instead of default namespace like `kube-system`.
- Charts deployed per user namespace and user no longer specifies the namespace explicitly. Namespaces need to be pre-created.
- Commands removed:
  - `helm reset`
  - `helm serve`
- The [Helm official Charts repository](https://github.com/helm/charts) is deprecated. It will be replaced by [Helm Hub](https://hub.helm.sh/) which is an  aggregator for different chart hosts.
- Release binaries are now hosted on `get.helm.sh`

The requirements for Helm 3 migrations are as follows:
- Helm 2 charts `install/template/--dry-run` as is
- Helm 2 and Helm 3 can work concurrently in the same cluster
- Helm 2 release and its history can be converted to Helm 3 release
- Helm 2 release and its history can be removed with and without removing the underlying Kubernetes resources
- Moving Helm 2 to 3 configuration 
- Ability to list releases by `--all-namespaces`
- Helm 2 data cleanup

Closes: https://github.com/helm/community/issues/67  

Tasks:

- [x] Migration documentation (https://github.com/helm/helm-www/pull/287 and #5582)
- [x] Tool for migrating Helm v2 releases in-place to v3 (#6154 and [2to3 plugin](https://github.com/helm/helm-2to3))
- [x] Tool for moving Helm 2 to 3 configuration ([2to3 plugin](https://github.com/helm/helm-2to3))
- [x] Tool for Helm 2 data cleanup ([2to3 plugin](https://github.com/helm/helm-2to3))

**Comment by hickeyma** [comment]
Investigations currently underway and progress shown in WIP doc PR #5582. In conjunction with the doc, there will be tools needed for the migration process.

One of these tools will be for migrating Helm 2 release objects to Helm 3. The tool was previously mentioned in https://github.com/helm/community/issues/67#issuecomment-448045222. 

From yesterdays dev weekly meeting, it’s suggested to write a proposal on the differences between Helm 2 and Helm 3’s object metadata and a how to migrate this data.

**Comment by hickeyma** [comment]
I will attach the proposal when ready.

**Comment by bacongobbler** [comment]
slating this for the 3.0.0 final release. It's a low priority to sort this out while we're still refactoring the internals, but we should certainly have a set of documentation available once 3.0 is ready to be released.

**Comment by hickeyma** [comment]
Sure @bacongobbler. Was distracted but working on this at the moment.

**Comment by hickeyma** [comment]
Proposal for migrating Helm v2 releases in-place to v3: https://github.com/helm/helm/issues/6154

**Comment by itowlson** [comment]
It looks like repositories added using Helm 2 are no longer found when using Helm 3 - is anybody else seeing this?  (I am on Windows and I'm using the default Helm 2 home for that OS.)

**Comment by hickeyma** [comment]
@itowlson A very point to raise. We haven't covered migration on local configuration from Helm 2 to Helm 3.  Thanks for raising this.

/cc @adamreese @bacongobbler @mattfarina @michelleN @prydonius @SlickNik @technosophos @thomastaylor312 @jdolitsky @henrynash @rimusz Should this capability be part of the  plugin?

**Comment by rimusz** [comment]
yes, it should be

**Comment by thomastaylor312** [comment]
@hickeyma Definitely think it should be as well. I think we should make sure it is a seamless as possible for people

**Comment by hickeyma** [comment]
Issue https://github.com/hickeyma/helm-2to3/issues/12 raised for moving Helkm 2 ro 3 configuration

**Comment by hickeyma** [comment]
I have added "Helm 2 uninstall and cleanup" to the migration requirements list in description. Should this be plugin functionality or a separate tool?

**Comment by rimusz** [comment]
it could be the plugin functionality, as why to make another tool

**Comment by hickeyma** [comment]
Issue https://github.com/helm/helm-2to3/issues/19 raised for cleanup. It is to be added to the  [helm-2to3](https://github.com/hickeyma/helm-2to3) plugin.

**Comment by hickeyma** [comment]
Helm v2 data cleanup functionality now in [2to3 plugin](https://github.com/helm/helm-2to3).

**Comment by hickeyma** [comment]
Migration document has now moved to https://github.com/helm/helm-www/pull/287

**Comment by hickeyma** [comment]
Closing as migration feature is now complete as follows:
- Helm 2 charts `install/template/--dry-run` supported as is
- List releases by `--all-namespaces`
- Helm 2 and Helm 3 can work concurrently in the same cluster
- [Migration document](https://v3.helm.sh/docs/intro/v2_v3_migration) published, for migrate in-place and manage both version information
- Helm [2to3](https://github.com/helm/helm-2to3) plugin in Helm Org, tool for migrate in-place

**Comment by scottrigby** [comment]
For anyone following this note:
>The Helm official Charts repository is deprecated. It will be replaced by Helm Hub which is an aggregator for different chart hosts.

See this issue for more detailed info: https://github.com/helm/charts/issues/16720#issuecomment-557250945
